sys/net/link_layer/ieee802154/security.c

```diff
@@ -0,0 +1,463 @@
+    ctx->frame_counter++;
+    return IEEE802154_SEC_OK;
+}
+
+int ieee802154_sec_decrypt_frame(ieee802154_sec_context_t *ctx,
+                                 uint8_t frame_size,
+                                 uint8_t *header,
+                                 uint8_t header_size,
+                                 uint8_t **auxiliary_header,
+                                 uint8_t *auxiliary_header_size,
+                                 uint8_t **payload,
+                                 uint8_t *payload_size,
+                                 uint8_t **mic,
+                                 uint8_t *mic_size,
+                                 const uint8_t *src_address)
+{
+    /* For non data frames (MAC commands, beacons) a_len would be larger.
+       ACKs are not encrypted. */
+    assert(*header & IEEE802154_FCF_TYPE_DATA);
+
+    ieee802154_aux_sec_t *aux = (ieee802154_aux_sec_t *)(header + header_size);
+    uint8_t security_level = _get_sec_level(aux->scf);
+    uint8_t key_mode = _get_key_id_mode(aux->scf);
+    uint8_t mac_size = _mac_size(security_level);
+
+    if (security_level == IEEE802154_SCF_SECLEVEL_NONE) {
+        *auxiliary_header = NULL;
+        *auxiliary_header_size = 0;
+        *payload = header + header_size;
+        *payload_size = frame_size - header_size;
+        *mic = NULL;
+        *mic_size = 0;
+        return IEEE802154_SEC_OK;
+    }
+    *auxiliary_header_size = _get_aux_hdr_size(security_level, key_mode);
+    *auxiliary_header = (uint8_t *)aux;
+    *payload_size = frame_size - header_size -
+                    *auxiliary_header_size - mac_size;
+    *payload = header + header_size + *auxiliary_header_size;
+    *mic_size = mac_size;
+    *mic = header + frame_size - mac_size;
+
+    uint8_t a_len = header_size + *auxiliary_header_size;
+    uint8_t c_len = *payload_size;
+
+    const uint8_t *key;
+    if (!(key = _get_decryption_key(ctx, header, header_size, aux))) {
+        return -IEEE802154_SEC_NO_KEY;
+    }
+    _set_key(ctx, key);
+
+    /* for reference, see the examples in the specification */
+    /* In the linux kernel: net/mac802154/llsec.c may also be an interesting reference */
+    /* better decrypt block by block, than holding
+       a large buffer because I had stack issues */
+    uint8_t block_buffer1[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+    uint8_t block_buffer2[IEEE802154_SEC_BLOCK_SIZE] = { 0 };
+    uint8_t *c = *payload;
+    uint8_t *mac = *mic;
+    uint8_t size, offset, blk;
+    ieee802154_ccm_block_t Ai;
+    Ai.flags = IEEE802154_CCM_FLAG(0, 2);
+    Ai.nonce.frame_counter = htonl(aux->fc);
+    /* A better implementation would check if the received frame counter is
+       greater then the frame counter that has previously been received from
+       the other endpoint. This is done to protect against replay attacks.
+       But we do not store this information because we also do not have
+       a proper key store, to avoid complexity on embedded devices. */
```

Comment:
let's add a TODO here so we can grep for such things :wink: 
If we have a L2 database we can add such things.